### PR TITLE
support checksum with asterisk at begin of filepath

### LINKED ISSUE #71999: get_url support checksum format where filename begins with asterisk
##### SUMMARY
Support checksum format where file path begins with asterisk.
Checksum entry is like:
`03a415f52b0eb13fe94476075bd13d98822dc619 *bionic-server-cloudimg-armhf.img`

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
get_url

##### ADDITIONAL INFORMATION
Need to use checksum file at https://cloud-images.ubuntu.com/bionic/current/SHA1SUMS
were the file path begins with asterisk.

```yaml
  - name: get image
    get_url:
      checksum: sha1:https://cloud-images.ubuntu.com/bionic/current/SHA1SUMS
      url: https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-amd64.img
      dest: /images/
```
Currently I have got an error message:
```
TASK [get image] ************************************************************************************************
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "attributes": null,
            "backup": null,
            "checksum": "sha1:https://cloud-images.ubuntu.com/bionic/current/SHA1SUMS",
            "client_cert": null,
            "client_key": null,
            "content": null,
            "delimiter": null,
            "dest": "/images/",
            "directory_mode": null,
            "follow": false,
            "force": false,
            "force_basic_auth": false,
            "group": null,
            "headers": null,
            "http_agent": "ansible-httpget",
            "mode": null,
            "owner": null,
            "regexp": null,
            "remote_src": null,
            "selevel": null,
            "serole": null,
            "setype": null,
            "seuser": null,
            "sha256sum": "",
            "src": null,
            "timeout": 10,
            "tmp_dest": null,
            "unsafe_writes": null,
            "url": "https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-amd64.img",
            "url_password": null,
            "url_username": null,
            "use_proxy": true,
            "validate_certs": true
        }
    },
    "msg": "Unable to find a checksum for file 'bionic-server-cloudimg-amd64.img' in 'https://cloud-images.ubuntu.com/bionic/current/SHA1SUMS'"
}
```

### PR DESCRIPTION
**Pull request by profiprog**
##### SUMMARY
fixes #71999 support checksum format where filename begins with asterisk

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
get_url

### CODE REVIEW

File: lib/ansible/modules/get_url.py [hunk 1]
```diff
@@ -512,7 +512,7 @@ def main():
 
             # Look through each line in the checksum file for a hash corresponding to
             # the filename in the url, returning the first hash that is found.
-            for cksum in (s for (s, f) in checksum_map if f.strip('./') == filename):
+            for cksum in (s for (s, f) in checksum_map if f.strip('./*') == filename):
```
Comment (bcoca) [inline lib/ansible/modules/get_url.py:515]: one issue, '*' can be a legal characther at the start/end of a file name

Comment (sivel) [reply lib/ansible/modules/get_url.py:515]: Yeah, I think we need to improve the logic, to not just strip. But instead normalize the filename by checking if `f[0]` is `*` and then only stripping the first char, while also using `os.path.basename` to strip things like `./`.

Maybe this instead:

```diff
diff --git a/lib/ansible/modules/get_url.py b/lib/ansible/modules/get_url.py
index 94d1d88dcb..574e4fbe63 100644
--- a/lib/ansible/modules/get_url.py
+++ b/lib/ansible/modules/get_url.py
@@ -514,7 +514,9 @@ def main():
             for line in lines:
                 parts = line.split(None, 1)
                 if len(parts) == 2:
-                    checksum_map.append((parts[0], parts[1]))
+                    if parts[1][0] in './*':
+                        parts[1] = parts[1][1:]
+                    checksum_map.append((parts[0], os.path.basename(parts[1])))
             filename = url_filename(url)

             # Look through each line in the checksum file for a hash corresponding to
```

**Review by sivel — CHANGES_REQUESTED** [review]

**Comment by sivel** [comment]
This will also require tests and a changelog entry.

**Comment by Baryczka** [comment]
Any chance to introduce this PR in the future? So far for example is not possible to download docker-compose using provided sha256 url file as it contains <hash> *docker-compose-<arch> format

**Comment by samdoran** [comment]
Superseded by #74502.

**Pull request closed without merging**